ntoskrnl/mm/section.c

```diff
@@ -987,45 +987,43 @@ FreeSegmentPage(PMM_SECTION_SEGMENT Segment, PLARGE_INTEGER Offset)
     MmReleasePageMemoryConsumer(MC_USER, Page);
 }
 
+_When_(OldIrql == MM_NOIRQL, _IRQL_requires_max_(DISPATCH_LEVEL))
+_When_(OldIrql == MM_NOIRQL, _Requires_lock_not_held_(MmPfnLock))
+_When_(OldIrql != MM_NOIRQL, _Requires_lock_held_(MmPfnLock))
+_When_(OldIrql != MM_NOIRQL, _Releases_lock_(MmPfnLock))
+_When_(OldIrql != MM_NOIRQL, _IRQL_restores_(OldIrql))
+_When_(OldIrql != MM_NOIRQL, _IRQL_requires_(DISPATCH_LEVEL))
 VOID
 NTAPI
 MmDereferenceSegmentWithLock(PMM_SECTION_SEGMENT Segment, KIRQL OldIrql)
 {
-    BOOLEAN HaveLock = FALSE;
-
     /* Lock the PFN lock because we mess around with SectionObjectPointers */
     if (OldIrql == MM_NOIRQL)
     {
-        HaveLock = TRUE;
         OldIrql = MiAcquirePfnLock();
```

Comment:
continued from #3798 
@tkreuzer says
> Not a fan of this change. I don't see how it fixes anything. It just makes the function behave more unintuitively. What if I don't want to release the lock after calling this function? I would prefer to have 2 functions: MmDereferenceSegmentWithPfnLockHeld (which would not lock) and MmDereferenceSegment, which would acquire the lock and then call this function. That's way cleaner than all this passing of magic values.
I'm generally against lock changes in any function that isn't either called "AcquireXLock" or "ReleaseXLock".

Reply:
The issue we have here is that we use the PFN lock to protect SectionObjectPointers, Section object refcount & Segment flags. So when the section reference reaches zero, we want to dereference the segment, check if we have to free the segment and mark the segment as "in delete" under the lock. But then we have to release the lock because we may have to tinker with the FS driver in order to flush data. and reacquire it to clear SectionObjectPointers.
I also would like to see a solution with `MmDereferenceSegmentWithPfnLockHeld` but this looks tricky with the current state of things.

Reply:
I still don't like it, but for the sake of getting this fixed asap, I guess we need to live with that for now.